### PR TITLE
Verify rpc request user against client cert

### PR DESCRIPTION
**Pull request by mberhault**
I would like a bit of feedback on the reflection in rpc/codec/server.go:userFromRequest.
Believe or not, this is the first time I'm using the reflect package.

Of course, comments on the rest of the code are more than welcome, keeping in mind that this is still a work in progress.

### CODE REVIEW

File: rpc/codec/server.go [hunk 1]
```diff
@@ -79,6 +92,72 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
 	return nil
 }
 
+// userFromRequest takes a request proto and attemps to access
+// request.RequestHeader.User.
+func (c *serverCodec) userFromRequest(request proto.Message) (string, error) {
```
Comment (tschottdorf) [inline rpc/codec/server.go]: can this method not take a `request proto.Request`?

File: rpc/codec/server.go [hunk 2]
```diff
@@ -79,6 +92,72 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
...
+	var requestedUser string
+
+	switch request := request.(type) {
+	case *cockroach_proto.PingRequest, *cockroach_proto.GossipRequest:
+		// Ping and Gossip requests do not have a header: require the node user.
+		requestedUser = security.NodeUser
+	case *message.EchoRequest, *message.ArithRequest:
+		// These are for testing purposes only.
+		return nil
+	default:
```
Comment (tschottdorf) [inline rpc/codec/server.go]: s/default/proto.Request/
and then a `default` case which errors out.

Comment (MBerhault) [inline rpc/codec/server.go]: oh yeah, much better. I wasn't aware of proto.Request.
This simplifies everything.

File: rpc/codec/server.go (hunk 1 above)
Comment (MBerhault) [inline rpc/codec/server.go]: removed function now that "extract user from header" is just 3 lines

File: rpc/codec/server.go [hunk 3]
```diff
@@ -24,11 +24,17 @@ package codec
 import (
 	"bufio"
 	"bytes"
+	"crypto/tls"
 	"fmt"
 	"io"
 	"net/rpc"
 
+	"github.com/cockroachdb/cockroach/base"
+	cockroach_proto "github.com/cockroachdb/cockroach/proto"
```
Comment (tschottdorf) [inline rpc/codec/server.go]: doesn't really concern your PR, but I've noticed how `etcd/raft` named its proto package `raftpb`, so if it's not mandated by style to have our proto package called `proto`, we could just rename both the `proto` folder and the contained package to `roachpb` (or even shorter, `pb`), which would allow us to always use the canonical import path.

Comment (MBerhault) [inline rpc/codec/server.go]: that would be much better, but yes, this is rather outside the scope of this PR.

File: rpc/codec/server.go [hunk 4]
```diff
@@ -79,6 +91,58 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
...
+		if header == nil {
+			return util.Errorf("missing header in request: %+v", request)
+		}
+		requestedUser = header.GetUser()
+		if len(requestedUser) == 0 {
+			return util.Errorf("missing User in request header: %+v", header)
+		}
+	case *cockroach_proto.PingRequest, *cockroach_proto.GossipRequest, *cockroach_proto.RaftMessageRequest:
+		// Ping and Gossip requests do not have a header: require the node user.
+		// TODO(marc): it may be worth it to make them implement Request.
```
Comment (tschottdorf) [inline rpc/codec/server.go]: yeah, probably worth having everything implement that to allow you to be agnostic here and below.

File: rpc/codec/server.go (hunk 3 above)
Comment (tschottdorf) [inline rpc/codec/server.go]: absolutely. @petermattis?

File: rpc/codec/server.go (hunk 4 above)
Comment (MBerhault) [inline rpc/codec/server.go]: yup. I started working on that, but ran into protoc issues.
At the very least, I want to remove all test requests.

File: rpc/codec/server.go (hunk 3 above)
Comment (petermattis) [inline rpc/codec/server.go]: We (err, @tamird) is looking into splitting up the `proto` package. There is no particular reason to have the protos separated from their usage. For example, the `MVCC` protos should live in `storage/engine`. We might have a base protos package which would make sense to name `roachpb` or `pb`, but hold off on doing any renaming yet until @tamird gets started with the split.

File: rpc/codec/server.go [hunk 5]
```diff
@@ -79,6 +91,58 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
...
+		// Request interface: this covers most requests.
+		header := request.Header()
+		if header == nil {
+			return util.Errorf("missing header in request: %+v", request)
+		}
+		requestedUser = header.GetUser()
+		if len(requestedUser) == 0 {
+			return util.Errorf("missing User in request header: %+v", header)
+		}
+	case *cockroach_proto.PingRequest, *cockroach_proto.GossipRequest, *cockroach_proto.RaftMessageRequest:
```
Comment (petermattis) [inline rpc/codec/server.go]: Mentioning specify protos feels awkward to me. Why not just have everything that isn't a `cockroach_proto.Request` default to not authenticated?

Comment (MBerhault) [inline rpc/codec/server.go]: until Ping/Gossip/Raft have a RequestHeader, I need to have them default to "desired user = node". otherwise, we would just fail all those requests.

File: rpc/codec/server.go [hunk 6]
```diff
@@ -79,6 +91,58 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
...
+// For requests with a RequestHeader, we compare the header.User against
+// the client certificate Subject.CommonName.
+// For other requests, we either allow all (test requests), or
+// require the "node user".
+func (c *serverCodec) authenticateRequest(request proto.Message) error {
+	var err error
+	var requestedUser string
+
+	switch request := request.(type) {
+	case cockroach_proto.Request:
```
Comment (petermattis) [inline rpc/codec/server.go]: Rather than use `cockroach_proto.Request` here, I think you could define your own more restrictive interface:

```
type userRequest interface {
  GetUser() string
}
```

Then you could fix the `{Echo,Arith}Request` messages by adding a user field. Ditto for `{Ping,Gossip,RaftMessage}Request`.

Comment (MBerhault) [inline rpc/codec/server.go]: Good point. I'll settle for that in the short term. Long-term, we need to see if there are more things we may need from the request header.

File: rpc/client_test.go [hunk 7]
```diff
@@ -36,13 +36,14 @@ func init() {
 func TestClientHeartbeat(t *testing.T) {
 	addr := util.CreateTestAddr("tcp")
 
+	// Heartbeats are node-to-node requests: use server context for both server and client.
```
Comment (bdarnell) [inline rpc/client_test.go]: Maybe server context should be named node context instead, since it can be used as a client too.

File: rpc/codec/message/arith.proto [hunk 8]
```diff
@@ -9,8 +9,9 @@ option go_package = "message";
 import "gogoproto/gogo.proto";
 
 message ArithRequest {
-	optional int32 a = 1 [(gogoproto.nullable) = false];
-	optional int32 b = 2 [(gogoproto.nullable) = false];
+  optional string user = 1 [(gogoproto.nullable) = false];
```
Comment (bdarnell) [inline rpc/codec/message/arith.proto]: Indentation is funky here, at least in github's diff view. Mixing tabs and spaces?

File: rpc/codec/message/echo.proto [hunk 9]
```diff
@@ -13,7 +13,8 @@ option (gogoproto.marshaler_all) = true;
 option (gogoproto.unmarshaler_all) = true;
 
 message EchoRequest {
-	optional string msg = 1 [(gogoproto.nullable) = false];
+  optional string user = 1 [(gogoproto.nullable) = false];
```
Comment (bdarnell) [inline rpc/codec/message/echo.proto]: Indention is off here too.

File: proto/api.go [hunk 10]
```diff
@@ -86,6 +86,12 @@ type Request interface {
 	flags() int
 }
 
+// RequestWithUser is an interface for RPC requests that have a "requested user".
+type RequestWithUser interface {
```
Comment (petermattis) [inline proto/api.go]: `UserRequest`? Note that there is no need to define this interface here. You could define it in rpc/codec right next to its use. (And name it `userRequest` or something like that).

File: rpc/codec/server.go [hunk 11]
```diff
@@ -79,6 +90,43 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
 	return nil
 }
 
+// authenticateRequest takes a request proto and attempts to authenticate it.
+// Requests need to implement RequestWithUser.
+// We compare the header.User against the client certificate Subject.CommonName.
+func (c *serverCodec) authenticateRequest(request proto.Message) error {
+	// RequestWithUser must be implemented.
+	requestWithUser, ok := request.(cockroach_proto.RequestWithUser)
```
Comment (petermattis) [inline rpc/codec/server.go]: As mentioned above, you can just define the interface here. Within the `authenticateRequest` function even:

```
  type requestWithUser interface {
    GetUser() string
  }
```

File: rpc/codec/server.go [hunk 12]
```diff
@@ -79,6 +90,43 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
 	return nil
 }
 
+// authenticateRequest takes a request proto and attempts to authenticate it.
+// Requests need to implement RequestWithUser.
+// We compare the header.User against the client certificate Subject.CommonName.
+func (c *serverCodec) authenticateRequest(request proto.Message) error {
```
Comment (petermattis) [inline rpc/codec/server.go]: Rather than performing the authentication here, perhaps serverCodec should contain a `PreHook` that could be invoked on a message before it is dispatched. That would keep the `rpc/codec` package free of dependencies of other cockroach packages.

**Comment by petermattis** [comment]
LGTM, though I think a small change could reduce dependency creep.

**Comment by bdarnell** [comment]
LGTM

File: rpc/codec/message/arith.proto (hunk 8 above)
Comment (MBerhault) [inline rpc/codec/message/arith.proto]: ah yes. how I wish I had gofmt for everything. Fixed.

File: rpc/codec/message/echo.proto (hunk 9 above)
Comment (MBerhault) [inline rpc/codec/message/echo.proto]: Done.

File: proto/api.go (hunk 10 above)
Comment (MBerhault) [inline proto/api.go]: I thought about that but figured it would be clearer to have it in the same place as Request and its implementers. I'm happy moving it though if you insist.

Comment (petermattis) [inline proto/api.go]: Moving it removes the dependency on the `proto` package. I like that.

Comment (MBerhault) [inline proto/api.go]: Good point. Will move it along with the rest of the code into a hook someplace (most likely security/)

File: rpc/client_test.go (hunk 7 above)
Comment (MBerhault) [inline rpc/client_test.go]: Ah, yes. I was pondering how to make those clearer. I'll do it in a separate PR since we have TestServerContext all over the place.

File: rpc/codec/server.go (hunk 12 above)
Comment (MBerhault) [inline rpc/codec/server.go]: I haven't changed it to a hook yet, just moved the certificate check to connection time (which is the right thing to do either way).
code still depends on security just to check user vs NodeUser (this should probably move).
I'll try to add the hook such that we can use it from the http server as well.

File: rpc/codec/message/arith.proto [hunk 13]
```diff
@@ -9,8 +9,9 @@ option go_package = "message";
 import "gogoproto/gogo.proto";
 
 message ArithRequest {
-	optional int32 a = 1 [(gogoproto.nullable) = false];
-	optional int32 b = 2 [(gogoproto.nullable) = false];
+	optional string user = 1 [(gogoproto.nullable) = false];
```
Comment (petermattis) [inline rpc/codec/message/arith.proto]: Similar to `GossipRequest`, you could add an `arith.go` package and manually implemented `ArithRequest.GetUser()` method.

File: rpc/codec/server.go [hunk 14]
```diff
@@ -79,6 +87,49 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
 	return nil
 }
 
+// UserRequest is an interface for RPC requests that have a "requested user".
+type UserRequest interface {
```
Comment (petermattis) [inline rpc/codec/server.go]: No need to export this type.

**Comment by spencerkimball** [comment]
Nice.

File: rpc/server.go [hunk 15]
```diff
@@ -88,24 +88,28 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 	if r.Method != "CONNECT" {
-		w.Header().Set(util.ContentTypeHeader, "text/plain; charset=utf-8")
-		w.WriteHeader(http.StatusMethodNotAllowed)
-		io.WriteString(w, "405 must CONNECT\n")
+		http.Error(w, "405 must CONNECT", http.StatusMethodNotAllowed)
 		return
 	}
+
+	// Construct an authentication hook for this security mode and TLS state.
+	authHook, err := security.AuthenticationHook(s.context.Insecure, r.TLS)
```
Comment (MBerhault) [inline rpc/server.go]: @petermattis: how about something like this?

File: rpc/codec/server.go [hunk 16]
```diff
@@ -37,6 +38,9 @@ type serverCodec struct {
 
 	methods []string
 
+	// Authentication hook. May be nil in tests.
+	authenticationHook func(string) error
```
Comment (petermattis) [inline rpc/codec/server.go]: The hook doesn't need to be authentication specific. Make the function take a `proto.Message` and have the proto perform a type assertion or type switch to extract the user.

Comment (MBerhault) [inline rpc/codec/server.go]: Made the hook take the entire proto.Message (see security/tls.go)

File: rpc/codec/server.go (hunk 14 above)
Comment (MBerhault) [inline rpc/codec/server.go]: This has moved to security/tls.go, inside the authentication method and unexported.

File: rpc/codec/message/arith.proto (hunk 13 above)
Comment (MBerhault) [inline rpc/codec/message/arith.proto]: Done.

File: rpc/codec/rpc_test.go [hunk 17]
```diff
@@ -112,7 +112,7 @@ func listenAndServeArithAndEchoService(network, addr string) (net.Addr, error) {
 				log.Infof("clients.Accept(): %v\n", err)
 				continue
 			}
-			go srv.ServeCodec(NewServerCodec(conn))
+			go srv.ServeCodec(NewServerCodec(conn, nil /* no authentication hook */))
```
Comment (petermattis) [inline rpc/codec/rpc_test.go]: These comments rotted: `s/authentication hook/request hook/g`

File: rpc/codec/server.go (hunk 14 above)
Comment (petermattis) [inline rpc/codec/server.go]: Looks like you didn't delete `UserRequest` here.

File: rpc/server.go (hunk 15 above)
Comment (petermattis) [inline rpc/server.go]: Yep. Looks good.

File: rpc/codec/rpc_test.go (hunk 17 above)
Comment (MBerhault) [inline rpc/codec/rpc_test.go]: separate PR on its way to fix that and extraneous UserRequest.
